### DISCUSSION TITLE
NoneType Problem

### DISCUSSION THREAD

**Discussion by hseynli** [MatthiasValvekens/pyHanko#198] [Q&A]
Hello. 
I want to sign pdf files, bu the code in below returns " Message='NoneType' object has no attribute 'get_signature_mechanism'"
What is the problem and how can I fix it?

`cms_signer = signers.SimpleSigner.load(
    'path/to/signer/key.pem', 'path/to/signer/cert.pem',
    ca_chain_files=('path/to/relevant/certs.pem',),
    key_passphrase=b'secret'
)

with open('input.pdf', 'rb') as doc:
    w = IncrementalPdfFileWriter(doc)
    out = signers.sign_pdf(
        w, signers.PdfSignatureMetadata(field_name='Signature1'),
        signer=cms_signer,
    )`

**Comment by MatthiasValvekens** [comment]
Check your logs, `SimpleSigner.load(...)` will log an error message and (IIRC) return `None` if it fails to load the key material---granted, that's not exactly good API design, but then again, signing with in-memory keys isn't really intended for production use anyhow.

My guess is that you copy-pasted the call to `.load(...)` from the documentation without changing the file paths to point to your key and certificate(s). ;)

**Reply by hseynli** [reply]
Thanks for your answer. But I must say we want to sign pdf files with PAdES signature using x509 sertificate.
I'd like to know can I use your library for sign pdf files (PAdES) using x509 sertificate?

**Comment by hseynli** [comment]
The code below throws an exception:

`
signer = signers.SimpleSigner.load_pkcs12(
    pfx_file='test.pem', passphrase=b'secret'
)

timestamper = timestamps.HTTPTimeStamper(
    url='http://tsa.example.com/timestampService'
)

signature_meta = signers.PdfSignatureMetadata(
    field_name='Signature', md_algorithm='sha256',
    # Mark the signature as a PAdES signature
    subfilter=SigSeedSubFilter.PADES,
    # We'll also need a validation context
    # to fetch & embed revocation info.
    validation_context=ValidationContext(allow_fetching=True),
    # Embed relevant OCSP responses / CRLs (PAdES-LT)
    embed_validation_info=True,
    # Tell pyHanko to put in an extra DocumentTimeStamp
    # to kick off the PAdES-LTA timestamp chain.
    use_pades_lta=True
)

with open('input.pdf', 'rb') as inf:
    w = IncrementalPdfFileWriter(inf)
    with open('output.pdf', 'wb') as outf:
        signers.sign_pdf(
            w, signature_meta=signature_meta, signer=signer,
            timestamper=timestamper, output=outf
        )
`
Exception: 'NoneType' object has no attribute 'get_signature_mechanism'
The test.pem file in the same directory where py files located.
input.pdf file also located in the same directory.

**Reply by MatthiasValvekens** [reply]
Hi, it appears that you're trying to load a PEM file as a PKCS#12 file. That won't work: either you have your cert and your private key as two separate files (PEM/DER-encoded), or they're stored together in one PKCS#12 file. If it's the former, use `SimpleSigner.load(...)`. If it's the latter, use `SimpleSigner.load_pkcs12(...)`.

**Reply by hseynli** [reply]
Thanks for your answer. But I must say we want to sign pdf files with PAdES signature using x509 sertificate.
I'd like to know can I use your library for sign pdf files (PAdES) using x509 sertificate?

**Reply by MatthiasValvekens** [reply]
Hi @hseynli, that's certainly possible, PAdES signing is part of the core feature set of pyHanko.

Were you able to resolve the problem you had? Just in case it wasn't clear: bear in mind that you always need the private key corresponding to your X.509 cert in order to actually generate a signature. Whether it's a PAdES signature or not doesn't matter.

**Reply by hseynli** [reply]
Hello. Thanks for your answer.
I have the code below. And I want to sign pdf files with pades using the code below:

`def hex_string_readable(bytes):
    return ["{:02X}".format(x) for x in bytes]

for store in ["MY"]:
    for cert, encoding, trust in ssl.enum_certificates(store):
        #certificate = x509.load_der_x509_certificate(cert, backend = default_backend())   # cryptography < 3.1 
        certificate = x509.load_der_x509_certificate(cert)   # cryptography >= 3.1 
        if hex(certificate.serial_number)[2:100] == "11e618d4956fb5eb000300029bb5":
            print(certificate.issuer, certificate.not_valid_after)
            print(certificate.serial_number)
            print("Serial Number HEX: ", hex(certificate.serial_number))
            print(certificate.subject)`

Could you help me please sign pdf files using this code?

**Reply by MatthiasValvekens** [reply]
Hi @hseynli,

Sorry for the delay. From what I can see, this code only loads certificates, but I don't see anything that loads the corresponding key. Which library provides that `enum_certificates` function?
